meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java

```diff
@@ -210,4 +222,87 @@ public static MetaClientException translateZkExceptionToMetaclientException(ZkEx
       return new MetaClientException(e);
     }
   }
+
+  /**
+   * This function translate and group Zk exception code to metaclient code.
+   * It currently includes all ZK code on 3.6.3.
+   */
+  public static MetaClientException.Code translateKeeperCodeToMetaClientCode(
+      KeeperException.Code zkCode) {
+    // TODO: add log to track ZK origional code.
+    switch (zkCode) {
+      case AUTHFAILED:
+      case SESSIONCLOSEDREQUIRESASLAUTH:
+      case INVALIDACL:
+        return MetaClientException.Code.AUTHFAILED;
+
+      case CONNECTIONLOSS:
+        return MetaClientException.Code.CONNECTIONLOSS;
+
+      case BADARGUMENTS:
+        return MetaClientException.Code.INVALIDARGUMENTS;
+
+      case BADVERSION:
+        return MetaClientException.Code.BADVERSION;
+
+      case NOAUTH:
+        return MetaClientException.Code.NOAUTH;
+
+      case NOWATCHER:
+        return MetaClientException.Code.INVALIDLISTENER;
+
+      case NOTEMPTY:
+        return MetaClientException.Code.NOTLEAFENTRY;
+
+      case NODEEXISTS:
+        return MetaClientException.Code.ENTRYEXISTS;
+
+      case SESSIONEXPIRED:
+      case SESSIONMOVED:
+      case UNKNOWNSESSION:
+        return MetaClientException.Code.SESSIONERROR;
+
+      case NONODE:
+        return MetaClientException.Code.NOSUCHENTRY;
+
+      case OPERATIONTIMEOUT:
+        return MetaClientException.Code.OPERATIONTIMEOUT;
+
+      case OK:
+        return MetaClientException.Code.OK;
+
+      case UNIMPLEMENTED:
+        return MetaClientException.Code.UNIMPLEMENTED;
+
+      case RUNTIMEINCONSISTENCY:
+      case DATAINCONSISTENCY:
+        return MetaClientException.Code.CONSISTENCYERROR;
+
+      case SYSTEMERROR:
+      case MARSHALLINGERROR:
+      case NEWCONFIGNOQUORUM:
+      case RECONFIGINPROGRESS:
+        return MetaClientException.Code.DBSYSTEMERROR;
+
+      case NOCHILDRENFOREPHEMERALS:
+      case INVALIDCALLBACK:
+      case NOTREADONLY:
+      case EPHEMERALONLOCALSESSION:
+      case RECONFIGDISABLED:
+        return MetaClientException.Code.DBUSERERROR;
+
+      /*
+       * APIERROR is ZooKeeper Code value separator. It is never thrown by ZK server,
+       * ZK error codes greater than its value are user or client errors and values less than
+       * this indicate a ZK server.
+       */
+      default:
+        if (zkCode.intValue() < KeeperException.Code.APIERROR.intValue()
+            & zkCode.intValue() >= KeeperException.Code.SYSTEMERROR.intValue()) {
+          return MetaClientException.Code.DBSYSTEMERROR;
+        } else {
+          return MetaClientException.Code.DBUSERERROR;
```

Comment:
A better practice when you have if with some return. You can remove "else clause":

   if (zkCode.intValue() < KeeperException.Code.APIERROR.intValue()
            & zkCode.intValue() >= KeeperException.Code.SYSTEMERROR.intValue()) {
          return MetaClientException.Code.DBSYSTEMERROR;
        } 
      return MetaClientException.Code.DBUSERERROR;

Reply:
TFTR Updated. Created another PR for formatting and renaming only.